README.md

```diff
@@ -286,45 +292,131 @@ type Action = {
+
+Verifications have three core concepts:
+
+1. A claim that includes a reference to a Farcaster account and the external entity. The claim can be hashed to create a unique identifier for each claim.
+2. A directional proof from the external entity that is authorized to make the claim, showing intent to connect it to the Farcaster account.
+3. A directional proof from the Farcaster account accepting the request to associate the claim with the Farcaster account.
+
+Verification claims are structured as `VerificationClaim` objects:
+
+```ts
+type VerificationClaim = {
+  externalAddressUri: URI;
+  account: number;
+};
+```
+
+Verifications are added via `VerificationAdd` messages and removed via `VerificationRemove` messages. Here are the types for the add and remove message bodies:
+
+```ts
+type VerificationAddBody = {
+  externalAddressUri: string;
+  claimHash: string;
+  externalSignature: string;
+  externalSignatureType: 'eip-191-0x45'; // Will support other types over time
+  schema: 'farcaster.xyz/schemas/v1/verification-add';
+};
+
+type VerificationRemoveBody = {
+  claimHash: string;
+  schema: 'farcaster.xyz/schemas/v1/verification-remove';
+};
+```
+
+Notice how `VerificationAddBody` does not include the full claim object, because the claim can be reconstructed from the `externalAddressUri` and `account` attributes of the add message.
+
+#### Message Validation
+
+The `envelope` of `VerificationAdd` and `VerificationRemove` messages is validated the [same as other message types](#message-validation).
+
+Here are rules for `VerificationAdd` messages:
+
+1. `schema` must be known
+2. `claimHash` must be present and match `hashFn(claim)`
+3. `externalSignatureType` must be known
+4. `externalSignature` must be valid according to custom rules depending on `externalSignatureType`, which are defined in subsections below
+
+Here are rules for `VerificationRemove` messages:
+
+1. `schema` must be known
+2. `claimHash` must be present
+
+#### Set Construction
+
+Verifications are stored in the Verifications Set, which is a modified [Last-Write-Wins-Element Set](<https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#LWW-Element-Set_(Last-Write-Wins-Element-Set)>). Valid and unremoved verifications are stored as `VerificationAdd` messages in the `adds` set. Removed verifications are stored as `VerificationRemove` messages in the `removes` set. Messages are indexed by `claimHash` and a particular claim cannot exist in both sets simultaneously.
+
+Verifications can be re-added once they've been removed, and conflicts are resolved using timestamp. Verification information like the external entity address is effectively hidden once a verification is removed, because the `VerificationRemove` message only contains the claim hash.
+
+When a verification add message `a` is received:
+
+1. If `a` exists in the `removes` set
+   1. If existing remove message is more recent, discard `a`
```

Comment:
what happens if the timestamps collide?

Reply:
Higher lexicographical hash wins, will update docs. I hadn't added that because in the current implementation with only one `externalSignatureType` we won't get to this point, but you're right we should add for future-proofing.

Reply:
One in the `removes` set wins when they have the same timestamp, will update docs